### PR TITLE
Feature/15028 add edit associated service

### PR DESCRIPTION
**Pull request by MichaelRobinson-BJSS**

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/AssociatedServicesController.cs [hunk 1]
```diff
@@ -55,9 +55,47 @@ public async Task<IActionResult> AssociatedServices(CatalogueItemId solutionId,
...
         {
-            throw new NotImplementedException("To be implemented in Story 15028");
+            var solution = await solutionsService.GetSolution(solutionId);
+            if (solution is null)
+                return BadRequest($"No Solution found for Id: {solutionId}");
+
+            return View(new AddAssociatedServiceModel(solution));
+        }
+
+        [HttpGet("edit-associated-service/{associatedServiceId}")]
```
Comment (bjss-colin) [inline src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/AssociatedServicesController.cs:67]: Not sure if we have an explicit URL convention. In some places it's more of a RESTful resource based implementation, like `{associatedServiceId}/edit-associated-service`.

File: tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            if (sessionStorage.ContainsKey(key))
+            {
+                value = sessionStorage[key];
+                return true;
+            }
+            else
+            {
+                value = null;
+                return false;
+            }
```
Comment (bjss-chris) [inline tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs:64]: `else` is unnecessary here, we could just do

```c#
value = null;
return false;
```

Also, might this work as an alternative implementation?

```c#
public bool TryGetValue(string key, out byte[] value)
{
    value = sessionStorage.GetValueOrDefault(key);
    return value is not null;
}
```

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs (hunk 2 above)
Comment (bjss-colin) [reply tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs:64]: Could go one step further as `Dictionary<TKey, TValue>` also has a `TryGetValue` method

```csharp
public bool TryGetValue(string key, out byte[] value) => sessionStorage.TryGetValue(key, out value);
```

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/AssociatedServicesController.cs (hunk 1 above)
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/AssociatedServicesController.cs:67]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs [hunk 3]
```diff
@@ -12,19 +12,19 @@ public static ClientApplication GetClientApplication(this Solution solution)
             if (string.IsNullOrWhiteSpace(solution.ClientApplication))
                 return new ClientApplication();
 
-            return JsonConvert.DeserializeObject<ClientApplication>(solution.ClientApplication);
+            return JsonSerializer.Deserialize<ClientApplication>(solution.ClientApplication);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs:15]: There are a lot of instances that this should be applied to. I was planning a separate PR to remove the remaining Newtonsoft references and I will fix this issue at the same time if that's ok.

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs [hunk 4]
```diff
@@ -12,19 +12,19 @@ public static ClientApplication GetClientApplication(this Solution solution)
...
         }
 
         public static string[] GetFeatures(this Solution solution) => string.IsNullOrWhiteSpace(solution.Features)
             ? System.Array.Empty<string>()
-            : JsonConvert.DeserializeObject<string[]>(solution.Features);
+            : JsonSerializer.Deserialize<string[]>(solution.Features);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs:20]: As above

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/OrderService.cs [hunk 5]
```diff
@@ -90,9 +90,7 @@ public Task<Order> GetOrderForStatusUpdate(CallOffId callOffId)
 
         public async Task<Order> CreateOrder(string description, string odsCode)
         {
-            var organisationId = (await organisationService.GetOrganisationByOdsCode(odsCode)).Id;
-
-            var orderingParty = await dbContext.Organisations.FindAsync(organisationId);
+            var orderingParty = await dbContext.Organisations.SingleAsync(x => x.OdsCode == odsCode);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/OrderService.cs:93]: Done. I only said you had almost cured me!

File: src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs [hunk 6]
```diff
@@ -12,19 +12,19 @@ public static ClientApplication GetClientApplication(this Solution solution)
...
 
         public static List<Integration> GetIntegrations(this Solution solution)
         {
             if (string.IsNullOrWhiteSpace(solution.Integrations))
                 return new List<Integration>();
 
-            return JsonConvert.DeserializeObject<List<Integration>>(solution.Integrations);
+            return JsonSerializer.Deserialize<List<Integration>>(solution.Integrations);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Framework/Extensions/SolutionExtensions.cs:27]: As above

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            await context.Orders.AddAsync(order);
+            await context.SaveChangesAsync();
+
+            await service.DeleteOrderItem(order.CallOffId, orderItem2.CatalogueItemId);
+
+            var updatedOrder = await context.Orders.AsAsyncEnumerable().FirstOrDefaultAsync();
+
+            updatedOrder.OrderItems.Should().Contain(x => x.CatalogueItemId == orderItem1.CatalogueItemId);
+            updatedOrder.OrderItems.Should().NotContain(x => x.CatalogueItemId == orderItem2.CatalogueItemId);
+            updatedOrder.OrderItems.Should().Contain(x => x.CatalogueItemId == orderItem3.CatalogueItemId);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs:54]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs (hunk 2 above)
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Test.Framework/SharedMocks/MockHttpSession.cs:64]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Models/AddAnOrganisationModelTests.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        [CommonAutoData]
+        public static void WithValidConstruction_PropertiesSetAsExpected(
+            Organisation organisation,
+            List<Organisation> availableOrganisations)
+        {
+            var actual = new AddAnOrganisationModel(organisation, availableOrganisations);
+
+            actual.Organisation.Should().Be(organisation);
+            actual.AvailableOrganisations.Should().BeEquivalentTo(availableOrganisations);
+            actual.SelectedOrganisation.Should().Be(default(int));
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.WebApp.UnitTests/Areas/Admin/Models/AddAnOrganisationModelTests.cs:22]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Constants/SelectListsTests.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            SelectLists.MemorySizes.Should().BeEquivalentTo(new List<SelectListItem>
+                {
+                    new SelectListItem { Text = "256MB", Value = "256MB" },
+                    new SelectListItem { Text = "512MB", Value = "512MB" },
+                    new SelectListItem { Text = "1GB", Value = "1GB" },
+                    new SelectListItem { Text = "2GB", Value = "2GB" },
+                    new SelectListItem { Text = "4GB", Value = "4GB" },
+                    new SelectListItem { Text = "8GB", Value = "8GB" },
+                    new SelectListItem { Text = "16GB or higher", Value = "16GB or higher" },
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Constants/SelectListsTests.cs:22]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Extensions/TypeExtensionsTests.cs [hunk 10]
```diff
@@ -8,13 +10,28 @@ public static class TypeExtensionsTests
...
+        {
+            typeof(TestControllerNoAttribute).AreaName().Should().BeNull();
         }
 
+        [Area("Solutions")]
         internal static class TestController
         {
         }
+
+        internal static class TestControllerNoAttribute
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Framework.UnitTests/Extensions/TypeExtensionsTests.cs:33]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using AutoFixture;
+using AutoFixture.AutoMoq;
+using AutoFixture.Idioms;
+using AutoFixture.Xunit2;
+using FluentAssertions;
+using Microsoft.EntityFrameworkCore;
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs:8]: No, but for some strange reason Visual Studio isn't pointing that one out to me.

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            order.AddOrUpdateOrderItem(orderItem1);
+            order.AddOrUpdateOrderItem(orderItem2);
+            order.AddOrUpdateOrderItem(orderItem3);
+
+            await context.Orders.AddAsync(order);
+            await context.SaveChangesAsync();
+
+            await service.DeleteOrderItem(order.CallOffId, orderItem2.CatalogueItemId);
+
+            var updatedOrder = await context.Orders.AsAsyncEnumerable().FirstOrDefaultAsync();
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs:50]: It was in due to a type clash, but that clash appears to be due to the 'using Microsoft.EntityFramework' which probably explains why VS was not highlighting that using statement as redundant. Removing that using statement has resolved the clash so I have removed AsAsyncEnumerable now.

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+            updatedOrder.OrderItems.Should().Contain(x => x.CatalogueItemId == orderItem1.CatalogueItemId);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs:52]: Done

File: tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            order.AddOrUpdateOrderItem(orderItem);
+
+            await context.Orders.AddAsync(order);
+            await context.SaveChangesAsync();
+
+            await service.DeleteOrderItem(order.CallOffId, orderItem.CatalogueItemId);
+
+            var updatedOrder = await context.Orders.AsAsyncEnumerable().FirstOrDefaultAsync();
+
+            updatedOrder.OrderItems.Should().NotContain(x => x.CatalogueItemId == orderItem.CatalogueItemId);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply tests/NHSD.GPIT.BuyingCatalogue.Services.UnitTests/Orders/OrderItemServiceTests.cs:77]: Done

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**